### ISSUE TITLE
[Bug] moreh_norm_backward returns all-zero gradients (and garbage values) for p=±inf, where the forward moreh_norm special-cases these orders

### ISSUE THREAD

**Issue by heroxcreed** [tenstorrent/tt-metal#56248]
# Issue draft — tt-metal #NEW (moreh norm_backward p=±inf)

**Title**: `[Bug] moreh_norm_backward returns all-zero gradients (and garbage values) for p=±inf, where the forward moreh_norm special-cases these orders`

### Summary

`ttnn.operations.moreh.norm_backward` accepts any float `p` — and the matching
forward `moreh_norm` special-cases `p = +inf`, `p = -inf` and `p = 0` in its
factories (`IS_ZERO` / `MINUS_INF` defines, e.g.
`moreh_norm_program_factory_nc_other.cpp:219-225`). The backward factory has
**no such special-casing**: it feeds `p` (and `p - 1`) straight into
`get_floored_p_and_decimal_and_p_is_negative()`
(`moreh_norm_backward_program_factory.cpp:37-44,119-121`), whose
`static_cast<uint32_t>(floored_p)` is undefined behavior for `±inf`
(`[conv.fpint]`), and whose `decimal = p - floor(p)` is `inf - inf = NaN`.

The result is silently wrong gradients:
- `p = +inf`: **every argmax element's gradient is returned as 0.0** (expected:
  `dy` broadcast to the argmax position — PyTorch semantics), and some non-max
  elements come back with garbage values up to `1.5e38` / `inf`.
- `p = -inf`: the argmin gradient is right, but **non-argmin elements keep a
  non-zero gradient** (expected 0), up to the full `dy` magnitude.

### Evidence (ttsim, Blackhole simulator, ttnn 0.78.0 wheel, bf16)

Minimal deterministic reproducer (seed 0), `x = randn(2, 64)`, `dy = randn(2, 1)`,
`p = +inf`, `dim = 1`:

```
row 0: argmax=59
  expected dx[argmax] = -0.4590   (dy = -0.4590)
  ttnn    dx[argmax] = +0.0000
  expected non-zero grads in row: 1;  ttnn non-zero: 0
row 1: argmax=36
  expected dx[argmax] = +0.3359   (dy = -0.3359)
  ttnn    dx[argmax] = +0.0000
  expected non-zero grads in row: 1;  ttnn non-zero: 0
```

Wider sweep (4 shapes × dims 0/1/2, both orders):

| p | shape, dim | max abs err | argmax grads lost | garbage > 1e30 |
|---|---|---|---|---|
| +inf | (4,100) d1 | 1.54e38 | 5 | 1 |
| +inf | (2,64) d1 | 1.13 | 2 | 0 |
| +inf | (4,100) d0 | 1.16e38 | 99 | 3 |
| +inf | (3,5,40) d2 | inf | 16 | 1 |
| -inf | (4,100) d1 | 1.41 | 0 (argmin right) | 0 |
| -inf | (2,64) d1 | 1.13 | 0 | 0 |
| -inf | (4,100) d0 | 3.41 | 0 | 0 |
| -inf | (3,5,40) d2 | 2.47 | 0 | 0 |

Expected gradients from `torch.linalg.vector_norm(x, ord=±inf, dim).backward(dy)`
(subgradient convention: gradient lands on the argmax/argmin element).

### Root cause

`moreh_norm_backward_program_factory.cpp`:
- L37-44 `get_floored_p_and_decimal_and_p_is_negative(float p)`:
  `floor(±inf)=±inf`, `decimal = ±inf ∓ inf = NaN`,
  `static_cast<uint32_t>(±inf)` = UB (on this target wraps to 0 or 2^32-1,
  hence the garbage exponents).
- L119-121: both `p` and `p - 1` go through it; the kernel then computes
  `|x|^(p-1)` and `1/y^p` via `power_tile_with_abs_x_to_cb` /
  `power_and_recip_tile_to_cb` (`moreh_common.hpp:979,1077`) with those
  meaningless exponent args.

For contrast, the forward factories test and special-case exactly these orders:
`moreh_norm_program_factory_{nc,h,w}_other.cpp` set `IS_ZERO` for `p == 0` and
`MINUS_INF` for `p == -inf`, and the kernels
(`moreh_norm_{nc,h,w}_kernel.cpp`) then use `binary_max_tile` /
`negative_tile` instead of the power ladder. `p = +inf` falls out of the default
path as abs+max. The backward side never got the same treatment.

### Notes / suggested fix direction

- The nightly test `test_moreh_norm_backward` (test_moreh_norm.py:574)
  parametrizes `p` only over `[2.0, 2.5, -2.5]` — the ±inf orders are exercised
  only in the *forward* tests, which is presumably why this never surfaced.
- Suggested: mirror the forward's dispatch. Host-side special-case in the
  backward factory for `p == ±inf` (and `p == 0`), emitting the sub-gradient
  directly: `dx = where(x == argmax_along_dim(|x|), dy, 0)` for +inf (and the
  argmin/-negated variant for -inf). On device this can reuse the same
  abs/max/negative tiles the forward already produces.
- Related but distinct issues: #55585 (zero-norm slices → NaN; PR #55598 adds a
  zero guard but does not touch the ±inf decomposition) and #55587 (the same
  floored-p decomposition is UB in `moreh_clip_grad_norm`). This report is the
  norm_backward ±inf mechanism specifically — verified on-device (simulator)
  rather than source-level only.

### Reproducer

```python
import torch, ttnn

torch.manual_seed(0)
device = ttnn.open_device(device_id=0)
D = lambda t: ttnn.from_torch(t.contiguous(), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)

x = torch.randn(2, 64).bfloat16()
dy = torch.randn(2, 1).bfloat16()

y = torch.linalg.vector_norm(x.float(), ord=float("inf"), dim=1, keepdim=True)
x64 = x.float().requires_grad_(True)
torch.linalg.vector_norm(x64, ord=float("inf"), dim=1, keepdim=True).backward(dy.float())
ref = x64.grad  # gradient on the argmax element only

got = ttnn.to_torch(ttnn.operations.moreh.norm_backward(
    D(x), D(y.bfloat16()), D(dy), p=float("inf"), dim=1)).float()

for r in range(2):
    amax = x.float().abs()[r].argmax().item()
    print(f"row {r} argmax={amax}: expected {ref[r, amax].item():+.4f}, got {got[r, amax].item():+.4f}")
```

Environment: tt-metal ttnn 0.78.0 wheel (commit of 2026-09-11 main is affected,
no inf handling in `moreh_norm_backward_program_factory.cpp`), Blackhole ttsim
simulator, bf16 tensors, TILE layout.

**Comment by heroxcreed** [comment]
Thanks for the confirmation.

One note on the verification gap: everything in the issue was produced on
the
Blackhole ttsim simulator at the device level (not source-level replay), so
a
fix can be verified end-to-end without a card — before/after tables against
torch autograd for both ±inf orders across dims/shapes, plus a
zero-regression
check over the currently-passing orders (p ∈ {2.0, 2.5, -2.5}).

I'd like to take this fix. Plan, mirroring the forward's dispatch:

1. Host side (moreh_norm_backward_program_factory.cpp): detect p == ±inf
   before get_floored_p_and_decimal_and_p_is_negative() ever runs, and emit
   dedicated compute defines — exactly how the forward factories already do
it
   (IS_ZERO / MINUS_INF in
moreh_norm_program_factory_{nc,h,w}_other.cpp:219-225).
2. Kernel side (moreh_norm_backward_kernel.cpp): ±inf paths implement the
   sub-gradient directly instead of the power ladder — dx = dy * sign(x) at
the
   argmax(|x|) element (0 elsewhere) for +inf; the argmin(|x|) variant for
   -inf — reusing the abs / binary_max_tile / negative_tile primitives the
   forward kernels already use.
3. Tests: extend test_moreh_norm_backward's p parametrization (today only
   [2.0, 2.5, -2.5], which is presumably why this never surfaced) with
   float("inf") / float("-inf") so the nightly covers it going forward.

p == 0 hits the same undefined decomposition in the backward factory; torch
does not define a gradient for ord=0, so I'd either reject it explicitly
with
TT_FATAL or special-case it — maintainer preference, happy to include
either.

Could a maintainer assign this to me? PR will follow with the ttsim
verification table in the description.

**Comment by heroxcreed** [comment]
Fix is up: #56334 (3 files, +184 lines).

**Approach** — mirrors the forward op's special-casing pattern (`IS_ZERO` / `MINUS_INF` in `moreh_norm_program_factory_{nc,h,w}_other.cpp`): the backward factory emits a `NORM_INF` define for `std::isinf(p)` instead of feeding ±inf into the floored-exponent path, and the kernel computes the exact sub-gradient with its existing bcast idioms:

```
dx = sign(x) * dy * eq(|x| - y, 0)
```

One branch serves both signs: `y` equals `max|x|` for `+inf` and `min|x|` for `-inf`, so the equality mask selects the arg-extreme support either way — no separate `-inf` variant needed.

**Verification** (Blackhole ttsim, bf16, torch autograd reference):
- ±inf matrix (1D / 2D W-dim COL bcast / 2D H-dim ROW bcast / 4D `[5,8,78,77]` over `dim=0..3`): all `max|err| = 0` (bf16-exact inputs); 4D randomized `uniform_(-1,1)` upstream grads: `max|err| ≤ 0.002`.
- Zero regression on the finite-order path: existing `p ∈ {2.0, 2.5, -2.5}` backward suite (138 cases) passes unchanged.
- New regression tests: `test_moreh_norm_backward_p_inf` (`p ∈ {inf, -inf}` × dims N/C/H/W × keepdim, per-group-unique `|x|` construction so the torch reference is unambiguous; keepdim=False follows the official `check_dim` skip rules).

**On the design questions I raised earlier:**
1. *Tie semantics* — measured against torch: when the arg-extremum of `|x|` is tied across T elements, torch splits `dy` evenly (`dy/T` each), while the eq-mask path gives every tie member the full `dy`. I documented this as a known limitation in the PR rather than implementing the cross-core tie-count reduction (that's a larger, separate piece of work). Tests are constructed tie-free so the divergence is explicit.
2. *`p = 0` backward* — torch itself returns zero gradients for `p = 0` on most inputs / raises on zeros in the input; I left `p = 0` behavior unchanged in this PR (it's out of scope for the ±inf fix). Happy to handle it separately if you want it in the same PR.

Repro from the issue now returns `dx = [0, 1, 0, 0]` (was `[-inf, 0, 0, 0]`), and the `-inf` case returns the argmin gradient (was all zeros).

**Comment by heroxcreed** [comment]
Hi @jberkowitzTT — I reported this and would like to take it. The fix is already up as PR #56334 (3 files, +184), with the reproduction and root-cause analysis in the issue body.
Could you assign the issue to me? I'm happy to work through review feedback.
